Dphan/add android

dorotaphanSiili opened the pull request:
- [x] Still missing generating code, this will be added as a part of this PR
- [ ] readme https://jira.it.epicgames.com/browse/APIGEAR-349

Closes #  https://jira.it.epicgames.com/browse/APIGEAR-275


<!-- 
Thanks for creating this pull request 🤗

Please make sure that the pull request is limited to one type (docs, feature, etc.) and keep it as small as possible. You can open multiple prs instead of opening a huge one.
-->

<!-- If this pull request closes an issue, please mention the issue number below -->
Closes # <!-- Issue # here -->

## 📑 Description
<!-- Add a brief description of the pr -->

<!-- You can also choose to add a list of changes and if they have been completed or not by using the markdown to-do list syntax
- [ ] Not Completed
- [x] Completed
-->

## ✅ Checks
<!-- Make sure your pr passes the CI checks and do check the following fields as needed - -->
- [ ] My pull request adheres to the code style of this project
- [ ] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [ ] All the tests have passed

## ℹ Additional Information
<!-- Any additional information like breaking changes, dependencies added, screenshots, comparisons between new and old behavior, etc. -->

dorotaphanSiili committed:
feat: add value for enum

in api we assume that enum has an int value, that is not a case in java.
The java enums are objects, which are set constants. Not narrowed to an
integer type.
To make them work within our ecosystem that includes other technologies,
where enums are integer based, I added a int field value and functions
to obtain the value for serialization, and create an enum based on the
integer.

This is part of adding an andorid-jni feature.

dorotaphanSiili committed:
chores: change the structure of files

it is not a good practice to have a modules names that have dot
Usually the dot separates submodules

This is part of adding the android-jni feature.

dorotaphanSiili committed:
feat(api): extend interface

add the async version of methods. When using with rpc we don't want to
block main thread, the rpc anyway assumes the answer will arrive at some
point later.
add the isReady property, which is also required when using rpc
mechanisms.

This is part of adding the android-jni feature.

dorotaphanSiili committed:
fix(api): add and remove event listner

the generated names were incorrect, that is fixed now.

This is part of adding the android-jni feature.

dorotaphanSiili committed:
feat(api): property change now has only new value

The reason for change is alignment with other techologies implementations

This is part of adding the android-jni feature.

dorotaphanSiili committed:
feat(api): split api file

the java convention (and ease of use) is that each class or struct has its own file.
The api was split to fulfill that practice.

This is part of adding the android-jni feature.

dorotaphanSiili committed:
feat(stub): add stub feature

adds an empty implementation of generated interface.
It fully supports handling properties (with the property change singal
emission).
It generates empty methods, to be filled with bussiness logic.
It may serve as an example or base implementation.

It allows easy implementing android service side when provideded as a
backend, hence all the helper file, facory, and service starter which
implicitely start the service.

This is part of adding the android-jni feature.

dorotaphanSiili committed:
feat(android): add messages and basic parcelable

Adds all the "topics" of messages (an enum).
Introduces a parcelable type that wraps generated structs.
This works only for structs with simple types.
The parcelable will be extended with handling complex types in later
tasks.

This is part of adding the android-jni feature.

dorotaphanSiili committed:
feat(android): add messenger to service

Adds handing then messages: calls proper methods on backend when serving
incoming messages and sends messages when notified by backend about
changes that must be forwarded to clients.

Added initial tests.

This is part of adding the android-jni feature.

dorotaphanSiili committed:
feat(adroid): Add client with project files

adds a client adapter, that implements the ServiceConnection and the api
interface.
A client is generated per api interface and connects with its Service
counterpart, that may be provided by diferent applications.
It requires the desired package that exposes the service to be given for
binding.
The service must be already started - the intent does not start it.
The client allows registering an event listener to get info about all
the changes that service informs about.
In that way it allows adding a jni version of that listener, which will
be added later.

This is part of adding the android-jni feature.

dorotaphanSiili committed:
feat: extend event listener

adding new event and a way to fire it to get info about object isReady state
change (e.g may be used for connection established or backend ready).

This is part of adding the android-jni feature.

dorotaphanSiili committed:
feat(example) add client example

adds an almost ready to start application for test purposes.
IMPORTANT it generates the functionality only for first interface.
That is for usability purpose, some tabs could be introduced for each
interface, but this make example more complex.
It genereates buttons to request some propety changes:
IMPORTANT the logic how to change it needs to be implemented by user,
similar for executing methods, the user must provide proper parameters.
It prints the result of events received from service: property changes,
singals receivied, method results.

This is part of adding the android-jni feature.

dorotaphanSiili committed:
feat(adnroid): sync initial state of service side

add message and handling that allow synchronizing the client state with
the service state on the connection established. The service sends all
the values of properties.

This is part of adding the android-jni feature.

dorotaphanSiili committed:
feat(android): support enums

add parcelable wrappers for enums and add proper handling of them in
service and client.
IMPORTANT: it is not yet handled if enum is part of struct.

This is part of adding the android-jni feature.

dorotaphanSiili committed:
feat(api): add copy ctor for structs

In java the structs (java objects) are passed by ref. The copy
constructor helps proper managing of lifetime when handling with
Parcelable objects.
This is preparatory commit to be used for helpers when handling
arrays of structures.

This is part of adding the android-jni feature.

dorotaphanSiili committed:
feat(android): add helpers for handlig arrays

Add functions for (de)serializing arrays of parcelable elements.

This is part of adding the android-jni feature.

dorotaphanSiili committed:
feat(android): handle arrays data types

Adds handling of arrays of all types: simple types, enums, structs.
adds some go templates for easier handling same code in many places.

Add handling of arrays of any type for messanger in service.
NOTE: using template makes the code not aligned

This is part of adding the android-jni feature.

dorotaphanSiili committed:
feat(example): add test service example skeleton

service uses a stub as the backend.
It is generated only for the first api interface.
User needs to fill the logic of changing properties and emitting
singals.
All the buttons and text views are already correctly connected.
It provides a buttons to start or stop service, change properties, emit
singals. The methods are also working, if the client sends a request it
is served, but the logic needs to be implemented in the stub i.e the
_impl module.

This is part of adding the android-jni feature.

dorotaphanSiili committed:
feat(api): add gradle project files

adds necessary gradle files to allow the project to be build with e.g.
android studio.
The gradle was chosen because of compatibility with unreal build system,
as the main purpose of all the changes in the series of commits is to
allow easy data change between unreal and android.
Most of the modules will be used with unreal to achieve that.

This is part of adding the android-jni feature.

dorotaphanSiili committed:
feat(android): add skeleton for service adapter

an android service is created for each interface.
It allows setting a backend that fullfills the interface.
The intention is to be albe to provide java-jni impl e.g. by unreal (but
it can bu also pure java implementation).
The backend is created by factory, so it is in same thread as the
service.
Note that the services are created in applications thread, not in a separate
ones, so having many interfaces and thus many services does not cause
the many threads overhead.
The messenger part is not yet there. It will be added with following
tasks.
The service must be started implicitely with
context.startService(theService) function.

This is part of adding the android-jni feature.

dorotaphanSiili committed:
feat(jni): add service backend allowing jni impl

adds the version of the backend that has some native functions.
This will be used with e.g. unreal template.
There are no build files as unreal has its own system.

This is part of adding the android-jni feature.

dorotaphanSiili committed:
feat(android): add android service project files

add gradle files with which it can be build with android studio or with
unreal.

This is part of adding the android-jni feature.

dorotaphanSiili committed:
feat(adnroid): extend tests for service

add tests that actually check if service properly reacts on receiving
messages and sends messages when needed.

This is part of adding the android-jni feature.

dorotaphanSiili committed:
feat(andorid): add tests for client

This is part of adding the android-jni feature.

dorotaphanSiili committed:
feat(jni): add client jni bridge

It is an event listener that has native methods that needs to be
implemented with cpp code.
It allows adding e.g. unreal client adapter.
There are no project files added as unreal provides its own build
system.

This is part of adding the android-jni feature.

dorotaphanSiili committed:
docs: update on implemented features

dorotaphanSiili committed:
fix: change module name filter

dorotaphanSiili committed:
feat: structs support complex and nested struct as data fields

dorotaphanSiili committed:
feat: comparison operators for struct

Structs have now implemented equals function and hashCode.

dorotaphanSiili committed:
feat: improve comparing complex types

use better comparing functions for structures and arrays.
Use implemented equals() for the strucutres.
Tests requires now filling the structures with some data other than
default, cause before setting new value and expected notification it is
checked for equality (not set in that case).
With new compaision that checks the content not addres of object it
needs to have some values.

dorotaphanSiili committed:
feat: add example app

Adds an app that uses composite build and builds all dependant modules.
The modules are in the diretory of rootproject.
Each module must provide a version.
The app just makes instances of all intefaces implementations:
stub and if android feature enabled also the client and service
versions.
Adds a task to run all unit tests.

dorotaphanSiili committed:
add goldenmaster

dorotaphanSiili committed:
feat: add github workflow

dorotaphanSiili committed:
fix(jni): fix unbind, to work even if called multiple times

dorotaphanSiili committed:
fix(jni): clean up the static members for service adapter

dorotaphanSiili committed:
feat: support imports and externs

dorotaphanSiili committed:
fix: add goldenmaster for imports, externs, ref if

dorotaphanSiili committed:
fix(android): set correct class loader on receive rpc resp

dorotaphanSiili committed:
fix: fix interfaces comparison

dorotaphanSiili committed:
chores: update test-apis

dorotaphanSiili committed:
fix: fix for ref interfaces tests

dorotaphanSiili committed:
fix: clean up log levels

dorotaphanSiili committed:
fix: fix string comparision

dorotaphanSiili committed:
fix: remove unused reference which may cause mem leak

dorotaphanSiili committed:
chores: make intention explicit on used threads

alwa-siili commented on templates/api/api.java.tpl:
```diff
@@ -8,10 +8,27 @@ public class {{Camel .Module.Name}} {
+  {{- range  $idx, $m :=.Members }}
+    {{- if $idx}}, {{ end -}}
+      {{Camel .Name}}({{.Value}})
+  {{- end }};
```
I assume this is handling the newline comma, using the implicit `int` to `bool` conversion, right?

So, in current form, it evaluates to

```java
public enum JavaEnum {
    @JsonProperty("1")
    Value1,
    @JsonProperty("2")
    Value2,
}
```

... while with this change, we change it to

```java
public enum JavaEnum {
    Value1
    , Value2
}
```

If my understanding is right, what is the reason of complicating this range template, if java supports trailing commas in `enum`s? Is this style of enum definition preferred / required by some android java standard?

If this style is not required / preferred, we could simplify this to:

```
  {{- range .Members }}
      {{Camel .Name}}({{.Value}}),
  {{- end }};
```

alwa-siili replied:
Nevermind, the `{{- if $idx}}, {{ end -}}` won't be added in the new line. Due to the `- `, all whitespace (including newline) is trimmed, so the comma will be placed after the previous enum variant.

alwa-siili commented on templates/api/api.java.tpl:
```diff
@@ -8,10 +8,27 @@ public class {{Camel .Module.Name}} {
+      for ({{Camel .Name}} e : values()) {
+          if (e.value == value) return e;
+      }
+      throw new IllegalArgumentException("Unknown code: " + code);
```
What is the `code`? It isn't declared anywhere. Shouldn't that be `value` instead?

alwa-siili replied:
This was fixed in later commits

alwa-siili commented on templates/api/api.java.tpl:
```diff
@@ -68,8 +68,10 @@ public class {{Camel .Module.Name}} {
     // methods
```
Please use more meaningful commit header message. This will help searching through the commits later.

Currently, it looks like this:

```bash
$ git show --oneline 24ff18a3926f44d081da9ca22eb438055fd8eb7f
24ff18a feat(api): extend interface
```

Instead, consider something like this

```
feat(api): extend interface to support asynchronous method invocations
```

alwa-siili commented on templates/api/api.java.tpl:
```diff
@@ -73,23 +73,23 @@ public class {{Camel .Module.Name}} {
     bool _isReady();
     // signal listeners
-    void addEventListener(I{{Camel .Name }}EventListener listener);
+    int addEventListener(I{{Camel .Name }}EventListener listener);
```
This change seems unrelated. It isn't mentioned anywhere in the commit message.

Please describe why `addEventListener` needs to return an `int`.

alwa-siili commented on templates/api/api.java.tpl:
```diff
@@ -73,23 +73,23 @@ public class {{Camel .Module.Name}} {
 
```
This change is unrelated to adding support for android. Instead it's just fixing a bad template.

This patchset it already huge. Changes related to fixing the state of the template should be extracted to a preparatory MR.

alwa-siili replied:
This is not going to be done.

alwa-siili commented on templates/api/api.java.tpl:
```diff
@@ -51,7 +51,7 @@ public class {{Camel .Module.Name}} {
   {{- range .Module.Interfaces }}
```
This commit is not related to supporting android, instead it is aligning a stale `template-java` with other templates.

This patchset is already huge, changes related to aligning `template-java` with other templates shouldn't be part of it.

Please extract changes aligning this template with other to a separate PR.

alwa-siili replied:
This is not going to be extracted

alwa-siili commented on templates/api/enum.java.tpl:
```diff
@@ -0,0 +1,30 @@
+    }
+  }
+
+}
```
```
\ No newline at end of file
```

alwa-siili commented on templates/api/enum.java.tpl:
```diff
@@ -0,0 +1,30 @@
+      for ({{Camel .Enum.Name}} e : values()) {
+          if (e.value == value) return e;
+      }
+      throw new IllegalArgumentException("Unknown code: " + code);
```
What is the `code`? Shouldn't that be `value`?

alwa-siili replied:
Outdated

alwa-siili commented on templates/api/enum.java.tpl:
```diff
@@ -0,0 +1,30 @@
+      }
+      throw new IllegalArgumentException("Unknown code: " + code);
+    }
+  }
```
All these lines are indented by 2 spaces. Is this necessary?

alwa-siili replied:
All "styling" changes will be addressed in future.

alwa-siili commented on templates/api/interface.java.tpl:
```diff
@@ -0,0 +1,33 @@
+
+import java.util.concurrent.CompletableFuture;
+
+
+  public interface I{{Camel .Interface.Name }} {
+    // properties
+  {{- range .Interface.Properties }}
+    void set{{Camel .Name}}({{javaParam "" .}});
+    {{javaReturn "" . }} get{{Camel .Name}}();
+    void fire{{Camel .Name}}Changed({{javaType "" .}} newValue);
+  {{ end }}
+    // methods
+  {{- range .Interface.Operations }}
+    {{javaReturn "" .Return}} {{camel .Name}}({{javaParams "" .Params}});
+    {{javaAsyncReturn "" .Return}} {{camel .Name}}Async({{javaParams "" .Params}});
+  {{- end }}
+   {{- range .Interface.Signals }}
+   public void fire{{Camel .Name}}({{javaParams "" .Params}});
+   {{- end }}
+    boolean _isReady();
+    // signal listeners
+    void addEventListener(I{{Camel .Interface.Name }}EventListener listener);
+    void removeEventListener(I{{Camel .Interface.Name }}EventListener listener);
+  }
```
These lines are indented by 2 spaces.

alwa-siili replied:
All "styling" changes will be addressed in future.

alwa-siili commented on templates/api/eventlistener.java.tpl:
```diff
@@ -0,0 +1,17 @@
+{{- range .Module.Enums }}
+import {{dot .Module.Name}}.{{dot .Module.Name}}_api.{{Camel .Name}};
+{{- end }}
+
+  public interface I{{Camel .Interface.Name }}EventListener {
+  {{- range .Interface.Properties }}
+    void on{{Camel .Name}}Changed({{javaType "" .}} newValue);
+  {{- end }}
+  {{- range .Interface.Signals }}
+    void on{{Camel .Name}}({{javaParams "" .Params}});
+  {{- end }}
+  }
```
These lines are indented by 2 spaces

alwa-siili replied:
All "styling" changes will be addressed in future.

alwa-siili commented on templates/api/eventlistener.java.tpl:
```diff
@@ -0,0 +1,17 @@
+
+  public interface I{{Camel .Interface.Name }}EventListener {
+  {{- range .Interface.Properties }}
+    void on{{Camel .Name}}Changed({{javaType "" .}} newValue);
```
What is your opinion on adding a separate filters for `javaType`, `javaParam(s)` and `java(Async)Return` with and without prefixes?

In the final code provided in this PR the prefix isn't specified anywhere:

```bash
template-java on dphan/add_android
❯ grep -RP "(?:java(?:Async)?Return|javaType|javaParams?) \"\"" | wc -l
112

template-java on dphan/add_android
❯ grep -RP "(?:java(?:Async)?Return|javaType|javaParams?) \"[^\"]" | wc -l
0
```

It seems, that versions of these filters without specifying the prefix, which currently is always empty would be beneficial to remove the clutter.

dorotaphanSiili replied:
that was our convention, although if not used it may be changed

alwa-siili commented on templates/api/struct.java.tpl:
```diff
@@ -0,0 +1,18 @@
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+//TODO imports - may need some struct from this or imported module
```
Do we keep track of dependencies between structs / interfaces / enums?

alwa-siili replied:
This is done in later commits.

alwa-siili commented on templates/api/struct.java.tpl:
```diff
@@ -0,0 +1,18 @@
+    @JsonProperty("{{snake .Name}}")
+    public {{javaType "" .}} {{camel .Name}};
+  {{- end }}
+  }
```
Indentation

alwa-siili replied:
All "styling" changes will be addressed in future.

alwa-siili commented on templates/api/abstract.java.tpl:
```diff
@@ -0,0 +1,43 @@
+      }
+    }
+  {{ end }}
+    
```
Unnecessary empty line

```suggestion
```

alwa-siili commented on templates/api/build.gradle.tpl:
```diff
@@ -0,0 +1,12 @@
+
+dependencies {
+    implementation 'com.fasterxml.jackson.core:jackson-annotations:2.17.0'
+}
```
```
\ No newline at end of file
```

alwa-siili replied:
All "styling" changes will be addressed in future.

alwa-siili commented on templates/gradle.properties:
```diff
@@ -0,0 +1,21 @@
+# Enables namespacing of each library's R class so that its R class includes only the
+# resources declared in the library itself and none from the library's dependencies,
+# thereby reducing the size of the R class for that library
+android.nonTransitiveRClass=true
```
```
\ No newline at end of file
```

alwa-siili replied:
All "styling" changes will be addressed in future.

alwa-siili commented on templates/api/enum.java.tpl:
```diff
@@ -0,0 +1,30 @@
+  public enum {{Camel .Enum.Name}} {
+  {{- range  $idx, $m :=.Enum.Members }}
+    {{- if $idx}}, {{ end -}}
+      @JsonProperty("{{.Value}}")
```
We just removed the `@JsonProperty` annotation from "enum variant" in

```
d1d5972 feat: add value for enum
```

Why are we reintroducing it here?

alwa-siili commented on templates/api/additions.gradle.tpl:
```diff
@@ -0,0 +1,7 @@
+}
+
+dependencies {
+    implementation 'com.fasterxml.jackson.core:jackson-annotations:2.17.0'
```
Since we provide a `libs.versions.toml` file, maybe we could specify the version of `com.fasterxml.jackson.core` there?

alwa-siili commented on templates/api/enum.java.tpl:
```diff
@@ -0,0 +1,30 @@
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+
+  public enum {{Camel .Enum.Name}} {
```
Let's keep the `package` and `include`s separated.

```suggestion
package {{dot .Module.Name}}.{{dot .Module.Name}}_api;

import com.fasterxml.jackson.annotation.JsonProperty;

public enum {{Camel .Enum.Name}} {
```

alwa-siili replied:
All "styling" changes will be addressed in future.

alwa-siili commented on templates/android/service/iservicefactory.java.tpl:
```diff
@@ -0,0 +1,7 @@
+import {{dot .Module.Name}}.{{dot .Module.Name}}_api.I{{Camel .Interface.Name }};
+
+
+public interface I{{Camel .Interface.Name}}ServiceFactory {
```
Let's split the `package` and `import` statements.

```suggestion
package {{camel .Module.Name}}.{{camel .Module.Name}}_android_service;

import {{dot .Module.Name}}.{{dot .Module.Name}}_api.I{{Camel .Interface.Name }};

public interface I{{Camel .Interface.Name}}ServiceFactory {
```

alwa-siili replied:
All "styling" changes will be addressed in future.

alwa-siili commented on templates/android/service/iservicefactory.java.tpl:
```diff
@@ -0,0 +1,7 @@
+
+public interface I{{Camel .Interface.Name}}ServiceFactory {
+    public  I{{Camel .Interface.Name }} getServiceInstance();
+}
```
We shouldn't use `getXYZ()` function in factory pattern. Instead a `createXYZ()` or `makeXYZ()` is better.

`getXYZ()` indicates a supplier pattern, rather than a factory pattern, where an object may not be created when the function is called. On the other hand, the factory always creates a new instance of the object.

However, considering the future implementation of this interface, which will be introduced in

```bash
50311c0 feat(stub): add stub feature
```

... we should rather use `IXYZServiceSupplier` name for the interface and `XYZServiceSupplier` for the class implementing it

```suggestion
public interface I{{Camel .Interface.Name}}ServiceSupplier {
    public  I{{Camel .Interface.Name }} getServiceInstance();
}
```

... and rename the file accordingly.

dorotaphanSiili replied:
renamed to Provider 🤔

dorotaphanSiili replied:
renamed to provider

alwa-siili commented on templates/android/service/serviceadapter.java.tpl:
```diff
@@ -0,0 +1,206 @@
+import {{dot .Module.Name}}.{{dot .Module.Name}}_api.I{{Camel .Interface.Name }};
+import {{dot .Module.Name}}.{{dot .Module.Name}}_api.Abstract{{Camel .Interface.Name}};
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+public class {{Camel .Interface.Name }}ServiceAdapter extends Service
```
Let's split the imports and class declaration.

```suggestion
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class {{Camel .Interface.Name }}ServiceAdapter extends Service
```

alwa-siili replied:
All "styling" changes will be addressed in future.

alwa-siili commented on templates/android/service/serviceadapter.java.tpl:
```diff
@@ -0,0 +1,206 @@
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+public class {{Camel .Interface.Name }}ServiceAdapter extends Service
+{
```
```suggestion
public class {{Camel .Interface.Name }}ServiceAdapter extends Service {
```

alwa-siili commented on templates/android/service/serviceadapter.java.tpl:
```diff
@@ -0,0 +1,206 @@
+	private static I{{Camel .Interface.Name}} mBackendService;
+	private static I{{Camel .Interface.Name}}ServiceFactory mServiceFactory;
+
+	//private final List<Message> mMessagesQueue = new ArrayList<>();
```
Please, don't commit commented out code. If it's unused remove it. If it's used in another commit, introduce it in the commit, which uses it.

alwa-siili replied:
All "styling" changes will be addressed in future.

alwa-siili commented on templates/android/service/serviceadapter.java.tpl:
```diff
@@ -0,0 +1,206 @@
+					}
+				}
+			}
+		}
```
This function isn't used anywhere in this commit. It would be a good idea to remove it from this patch and reintroduce it in commit which uses it. This way, the function will provide more concept for the future commit.

alwa-siili commented on templates/android/service/serviceadapter.java.tpl:
```diff
@@ -0,0 +1,206 @@
+			Log.i(TAG, "LIFECYCLE: IncomingHandler(finalize)");
+		}
+
+		private void addClientActivity(Messenger serviceReply, String connectionID)
+		{
+			if (serviceReply != null)
+			{
+				mActivityClients.put(connectionID, serviceReply);
+				Log.i(TAG, "Register event listener with connectionID = " + connectionID);
+			}
+		}
+
+		private void removeClientActivity(String connectionID)
+		{
+			mActivityClients.remove(connectionID);
+			Log.i(TAG, "UnRegister event listener with connectionID = " + connectionID);
+		}
```
These functions aren't used in this commit. It would be a good idea to remove them and reintroduce them in the commit, which actually uses them. This way, these functions provide additional context for the future commit.

alwa-siili commented on templates/android/service/serviceadapter.java.tpl:
```diff
@@ -0,0 +1,206 @@
+	class IncomingHandler extends Handler implements I{{Camel .Interface.Name }}EventListener
+	{
+		private final Service mApplicationContext;
+		private final ConcurrentHashMap<String, Messenger> mActivityClients = new ConcurrentHashMap<>();
```
This variable is only accessed in functions, which aren't used in this commit. It would be a good idea to remove it and reintroduce it in the commit which uses the functions.

alwa-siili commented on templates/android/service/serviceadapter.java.tpl:
```diff
@@ -0,0 +1,206 @@
+	@Override
+	public int onStartCommand(Intent intent, int flags, int startId)
+	{
+		Log.i(TAG, "LIFECYCLE: {{Camel .Interface.Name }}Service::onStartCommand called. context = " + this +
+				", startID=" + startId);
```
Consider formatting the log lines

```suggestion
		Log.i(TAG,
            "LIFECYCLE: {{Camel .Interface.Name }}Service::onStartCommand called. " +
            String.format("context = %s, startId = %s", this, startId));
```

Similarly in other logs.

alwa-siili replied:
All "styling" changes will be addressed in future.

alwa-siili commented on templates/android/service/serviceadapter.java.tpl:
```diff
@@ -0,0 +1,206 @@
+	{
+		super.onDestroy();
+		
+		Log.i(TAG, "LIFECYCLE: onDestroy({{Camel .Interface.Name }}Service) - proc = " + ", mMessenger = " + mMessenger
```
The value of `proc` is missing in the log message. Same below

alwa-siili commented on templates/android/service/serviceadapter.java.tpl:
```diff
@@ -0,0 +1,206 @@
+		return super.onUnbind(intent);
+	}
+
+	private static String Name(Context context)
+	{
+		return context.getPackageName() + ",context=" + context;
+	}
```
Unused function.

alwa-siili commented on templates/android/service/serviceadapter.java.tpl:
```diff
@@ -0,0 +1,206 @@
+
+	//private final List<Message> mMessagesQueue = new ArrayList<>();
+
+	public {{Camel .Interface.Name }}ServiceAdapter()
+	{
+	}
```
1. Shouldn't we call `super()` here?
2. Isn't this the same as the default constructor?

dorotaphanSiili replied:
yes, should be removed, probably a leftover after it did something

alwa-siili commented on templates/android/service/serviceadapter.java.tpl:
```diff
@@ -0,0 +1,206 @@
+	 */
+	class IncomingHandler extends Handler implements I{{Camel .Interface.Name }}EventListener
+	{
+		private final Service mApplicationContext;
```
An instance of `IncomingHandler` is held static in `ServiceAdapter`, while it keeps a reference to `context`, which is an instance of `ServiceAdapter`, which created it. Is this safe? This may prevent the `Service` from being removed after `onDestroy` until the next service is created and `onCreate` is called.

I would propose to use a `java.lang.ref.WeakReference<Service>` here

```suggestion
		private final WeakReference<Service> mWeakApplicationContext;
```

... and then accessing it as

```java
Service context = mWeakApplicationContext.get();
if (context != null)
{
    // do something with context
}
```

This way we're not preventing GC from removing the `ServiceAdapter`, which created this (static) instance of `IncomingHandler`

alwa-siili replied:
The `mApplicationContext` member is removed in later commits.

alwa-siili commented on templates/android/service/AndroidManifest.xml.tpl:
```diff
@@ -0,0 +1,23 @@
+		</service>
+        {{- end }}
+    </application>
+</manifest>
```
```
\ No newline at end of file
```

alwa-siili replied:
All "styling" changes will be addressed in future.

alwa-siili commented on templates/android/service/additions.gradle.tpl:
```diff
@@ -0,0 +1,18 @@
+android {
+    namespace '{{camel .Module.Name}}.{{camel .Module.Name}}_android_service'
+
+	tasks.register('packageSources', Jar) {
+		archiveClassifier.set('sources')
+		from android.sourceSets.main.java.srcDirs
+		include '**/*.java'
+	}
+}
+
+dependencies {
+	implementation 'androidx.appcompat:appcompat:1.2.0'
+    implementation project(':{{camel .Module.Name}}_api')
+    testImplementation 'junit:junit:4.13.2'
+    testImplementation 'org.robolectric:robolectric:4.10.3'
+    testImplementation 'org.mockito:mockito-core:5.12.0'
+    testImplementation 'org.mockito:mockito-inline:5.2.0'
+}
```
Mixing tabs and spaces. Please select one and keep it consistent.

alwa-siili replied:
All "styling" changes will be addressed in future.

alwa-siili commented on templates/android/service/build.gradle.tpl:
```diff
@@ -0,0 +1,27 @@
+
+    defaultConfig {
+        minSdk 34
+		targetSdk 34
```
In unreal we target an older SDK (33). Should we lower the `minSdk` here as well?

alwa-siili replied:
We did in following commits...

alwa-siili commented on rules.yaml:
```diff
@@ -36,7 +36,26 @@ features:
         documents:
```
Typos in commit message:
* `singal` -> `signal`
* `implicitely` -> `implicitly`

alwa-siili commented on templates/android/service/implserviceprovider.java.tpl:
```diff
@@ -0,0 +1,81 @@
+ */
+public class {{Camel .Interface.Name}}ServiceFactory extends HandlerThread implements I{{Camel .Interface.Name }}ServiceFactory
+{
+	private {{Camel .Interface.Name}}Service m_Service;
```
Modern java and android ditched prefixes (`m_`, `m`) of class members. Should we follow that?

If not, then let's at least choose one prefix for all java code. We've used `m` prefix in previous commits. Now, we introduce `m_` prefix. Let's make it coherent.

alwa-siili commented on templates/android/service/implservicefactory.java.tpl:
```diff
@@ -0,0 +1,81 @@
+	{
+		synchronized (this)
+		{
+			Log.w("UE", "LIFECYCLE: {{Camel .Interface.Name}}ServiceFactory::onDestroy() - stop instance thread, service = " + m_Service);
```
1. Let's use a `TAG` instead.
2. Why are we using `UE` as a tag? It implies, that this class is intended to be used by Unreal Engine exclusively, which isn't true, because we could use it with plain android application as well.
3. Why is it a warning it should be either a `debug` or `info`?
4. Formatting. Do we follow any max line length convention? Google and Java uses 100 characters per line, with exceptions. We could at least loosely follow their convention.

```suggestion
			Log.w("UE",
                "LIFECYCLE: {{Camel .Interface.Name}}ServiceFactory::onDestroy() - stop instance thread, " +
                "service = " + m_Service);
```

Same below.

alwa-siili replied:
All "styling" changes will be addressed in future.

alwa-siili commented on templates/android/service/implservicefactory.java.tpl:
```diff
@@ -0,0 +1,81 @@
+		}
+	}
+
+	public synchronized  Abstract{{Camel .Interface.Name }} getServiceInstance()
```
Unnecessary double space

```suggestion
	public synchronized Abstract{{Camel .Interface.Name }} getServiceInstance()
```

alwa-siili commented on templates/android/service/implservicefactory.java.tpl:
```diff
@@ -0,0 +1,81 @@
+ * @hide
+ */
+public class {{Camel .Interface.Name}}ServiceFactory extends HandlerThread implements I{{Camel .Interface.Name }}ServiceFactory
+{
```
What is the convention for the opening brace? In some cases, we add it in the same line as `class`, in the other we add it the following line. Let's choose some convention and use it everywhere.

[Google Java Style](https://google.github.io/styleguide/javaguide.html) (and Oracle) states, that all opening braces should be on the end of the line (not in the newline). Should we follow that and apply it to all of our java code?

Note, that Google also enforces 2 line indentations, so we should consider, whether we want to follow this convention or not.

@jryannel @dorotaphanSiili @w4bremer what's your opinion?

alwa-siili commented on templates/android/service/implservicefactory.java.tpl:
```diff
@@ -0,0 +1,81 @@
+
+	private {{Camel .Interface.Name}}ServiceFactory()
+	{
+		super("EngineFactory");
```
Why are we using `EngineFactory` as the name, and not `{{Camel .Interface.Name}}ServiceFactory`?

dorotaphanSiili replied:
not present there anymore

alwa-siili commented on templates/android/service/implservicefactory.java.tpl:
```diff
@@ -0,0 +1,81 @@
+		if (m_Service == null)
+		{
+			m_Service = new {{Camel .Interface.Name}}Service();
+			Log.d("UE", "LIFECYCLE: EngineFactory::GetInstance(with OBBFilename) - CREATED new {{Camel .Interface.Name}}Service");
```
Why are we using `EngineFactory` here, and not `{{Camel .Interface.Name}}ServiceFactory` like above?

Same below.

dorotaphanSiili replied:
not present there anymore

alwa-siili commented on templates/android/service/implservicestarter.java.tpl:
```diff
@@ -0,0 +1,42 @@
+package {{camel .Module.Name}}.{{camel .Module.Name}}_android_service;
```
Missing copyright.

alwa-siili replied:
All "styling" changes will be addressed in future.

alwa-siili commented on templates/android/service/implservicestarter.java.tpl:
```diff
@@ -0,0 +1,42 @@
+    static Intent androidService = null;
+    private static final String TAG = "{{Camel .Interface.Name }}Starter";
+
+
```
Unnecessary empty lines

```suggestion
```

alwa-siili commented on templates/android/service/implservicestarter.java.tpl:
```diff
@@ -0,0 +1,42 @@
+
+
+
+    public static I{{Camel .Interface.Name }} start(Context context) {
```
Mixing opening brace position.

alwa-siili commented on templates/stub/implservice.java.tpl:
```diff
@@ -0,0 +1,99 @@
+    }
+    {{- end }}
+
+}
```
```
\ No newline at end of file
```

alwa-siili commented on templates/settings.gradle.tpl:
```diff
@@ -20,6 +20,12 @@ dependencyResolutionManagement {
 }
 
 rootProject.name = "{{camel .Module.Name}}"
+{{- if .Features.android }}
 include ':{{camel .Module.Name}}_android_service'
+{{- end -}}
+{{- if .Features.stubs }}
+include ':{{camel .Module.Name}}_impl'
+{{- end -}}
+{{- if .Features.api }}
 include ':{{camel .Module.Name}}_api'
-
+{{- end -}}
```
How does features work? If I select `stubs` without `api` will it apigear-cli automatically select the `api` as well? If not, the code will be incomplete.

alwa-siili replied:
Feature dependencies are specified in the `rules.yaml` file. When a feature `stub` is selected, feature `api` will automatically be selected as well.

alwa-siili commented on templates/stub/additions.gradle.tpl:
```diff
@@ -0,0 +1,18 @@
+android {
```
What actually is the file `additions.gradle` for? Is it something related to GooglePlay? Is this automatically used by gradle/gradlew/something else? I cannot find any meaningful information regarding `additions.gradle` files.

alwa-siili replied:
This is a build file used by Unreal Engine (instead of `build.gradle` file used by Android Studio).

alwa-siili commented on templates/stub/implservice.java.tpl:
```diff
@@ -0,0 +1,99 @@
+    private static final ExecutorService executor = Executors.newFixedThreadPool(1);
+
+    {{- range .Interface.Properties }}
+    private {{javaReturn "" .}} m_{{javaVar  .}} = {{ javaDefault "" . }};
```
`javaType` would be appropriate here

```suggestion
    private {{javaType "" .}} m_{{javaVar  .}} = {{ javaDefault "" . }};
```

alwa-siili replied:
It works either way, they are both the same.

alwa-siili commented on templates/android/messenger/messages.java.tpl:
```diff
@@ -0,0 +1,48 @@
+package {{camel .Module.Name}}.{{camel .Module.Name}}_android_messenger;
+
+{{ define "inc" }}{{ len (printf "%*s " . "") }}{{ end -}}
```
This defined function isn't used anywhere. Instead the body of the function is copied in each `msgNum`.

alwa-siili commented on templates/android/service/serviceadapter.java.tpl:
```diff
@@ -162,10 +164,102 @@ public class {{Camel .Interface.Name }}ServiceAdapter extends Service
+					{{- if not .IsPrimitive }}
+					data.setClassLoader({{Camel .Type}}Parcelable.class.getClassLoader());
+					{{- end }}
+					{{- end }}
```
This doesn't make sense, a bundle can have only a single class loader at a time.

~~However, since we're using `T Bundle.getParcelable (String key, Class<T> clazz)`, there is no need to set the class loader anymore. However, using the new `getParcelable` method implies, that the target device must support API33 (Android 13 or newer).~~

alwa-siili replied:
Outdated anyway.

alwa-siili commented on templates/android/service/serviceadapter.java.tpl:
```diff
@@ -162,10 +164,102 @@ public class {{Camel .Interface.Name }}ServiceAdapter extends Service
+
+					{{- if .Return.IsPrimitive }}
+					{{ if not .Return.IsVoid }}{{javaReturn "" .Return}} result = {{- end }}mBackendService.{{camel .Name}}({{javaVars .Params}});
+					{{- else }}
```
The check `{{ if not .Return.IsVoid }}` doesn't make sense, since it should always evaluate to true.

According to https://github.com/apigear-io/cli/blob/ce5795bc952de9687177aac464de7b9063f48d5a/pkg/model/schema.go#L58, type `void` is not primitive.

alwa-siili replied:
Outdated

alwa-siili commented on templates/android/service/serviceadapter.java.tpl:
```diff
@@ -162,10 +164,102 @@ public class {{Camel .Interface.Name }}ServiceAdapter extends Service
+					{{ if not .Return.IsVoid }}{{javaReturn "" .Return}} result = {{- end }}mBackendService.{{camel .Name}}({{javaVars .Params}});
+					{{- else }}
+					{{ if not .Return.IsVoid }}{{javaReturn "" .Return}} result = {{- end }} mBackendService.{{camel .Name}}({{javaVars .Params}});
+					{{ if not .Return.IsVoid }}{{javaReturn "" .Return}}
```
Should the `{{javaReturn "" .Return}}` be here? It doesn't seem to make sense here.

alwa-siili replied:
Outdated

alwa-siili commented on templates/android/service/serviceadapter.java.tpl:
```diff
@@ -162,10 +164,102 @@ public class {{Camel .Interface.Name }}ServiceAdapter extends Service
+					{{- if .Return.IsPrimitive }}
+					{{ if not .Return.IsVoid }}{{javaReturn "" .Return}} result = {{- end }}mBackendService.{{camel .Name}}({{javaVars .Params}});
+					{{- else }}
+					{{ if not .Return.IsVoid }}{{javaReturn "" .Return}} result = {{- end }} mBackendService.{{camel .Name}}({{javaVars .Params}});
```
This should be `dataResult` instead.

```suggestion
					{{ if not .Return.IsVoid }}{{javaReturn "" .Return}} dataResult = {{- end }} mBackendService.{{camel .Name}}({{javaVars .Params}});
```

alwa-siili replied:
Outdated

alwa-siili commented on templates/android/client/client.java.tpl:
```diff
@@ -0,0 +1,386 @@
+    {{- end}}
+
+
+	public {{Camel .Interface.Name }}Client(Context applicationContext, String connectionId)
+	{
+        assert (applicationContext != null);
```
Since this code is strictly an android part of the project, consider annotating `Context` with [@NonNull](https://developer.android.com/studio/write/annotations#adding-nullness).

This will generate compile time warnings when using android studio if provided context could be null.

alwa-siili commented on templates/android/client/client.java.tpl:
```diff
@@ -0,0 +1,386 @@
+
+        if (connectionId.isEmpty())
+        {
+            mConnectionId = UUID.randomUUID().toString();
```
If we have multiple clients, possibly from different packages / devices, we can't guarantee that strings created by `UUID.randomUUID()` will always be unique.

Perhaps a better approach would be to pass the responsibility of creating a connection ID to the service, where a client would "ping" the service and request a new connection ID. This way, we can guarantee that connection IDs used by a single service instance will never be duplicated.

Anyway, that's not an issue for now, rather something to consider later.

alwa-siili commented on templates/android/client/client.java.tpl:
```diff
@@ -0,0 +1,386 @@
+        assert (applicationContext != null);
+        mApplicationContext = applicationContext;
+
+        if (connectionId.isEmpty())
+        {
+            mConnectionId = UUID.randomUUID().toString();
+        }
+        else
+        {
+            mConnectionId = connectionId;
+        }
```
We should prefer using a ternary operator in such cases, and use `if` blocks when we've got alternative logic depending on some condition.

In case of ternary operator, I would also invert the condition in this case, checking if the `connectionId` is not empty.

```suggestion
        mApplicationContext = !connectionId.isEmpty() ? connectionId : UUID.randomUUID().toString();
```

alwa-siili commented on templates/android/client/client.java.tpl:
```diff
@@ -0,0 +1,386 @@
+    public boolean bindToService(String packageName)
+    {
+        Intent intent = new Intent();
+        intent.setClassName(packageName, "{{camel .Module.Name}}.{{camel .Module.Name}}_android_service.{{Camel .Interface.Name }}ServiceAdapter");
```
Since we're always building service side and client side during android build, maybe we could add a dependency on the service to the client? Then we could import the `*ServiceAdapter` and use it here as `*ServiceAdapter.class`.

Currently, if we introduce changes to the `*ServiceAdapter` name or it's module name, we need to remember to apply changes in this string as well, otherwise we'll have a bug. Using `*ServiceAdapter.class` we would receive a compilation error in such cases, which is much better than runtime bug.

alwa-siili commented on templates/android/client/client.java.tpl:
```diff
@@ -0,0 +1,386 @@
+
+    private void requestRegisterClient()
+    {
+        if (mClientMessenger != null)
+        {
+            Message msg = Message.obtain(null, {{Camel .Interface.Name}}MessageType.REGISTER_CLIENT.ordinal());
+            msg.replyTo = mClientMessenger;
+            msg.getData().putString("connectionID", mConnectionId);
+            mClientHandler.sendToService(msg);
+        }
```
This will work only after the first binding to the service. Once a service is disconnected (because it crashed or was killed), we handle it in `onServiceDisconnected`. Then we null the `mClientMessager`.

However, until we unbind from the service, we may receive `onServiceConnected` again (when the service reconnects automatically). Then we'll call `requestRegisterClient`, but it will be a noop, since `mClientMessenger` is only created once in the constructor and never again.

Is this intentional?

dorotaphanSiili replied:
fixed

alwa-siili commented on templates/android/client/client.java.tpl:
```diff
@@ -0,0 +1,386 @@
+			    case SET_{{Camel .Name}}:
+			    {
+				    Bundle data = msg.getData();
+			    {{- .IsPrimitive }}
```
Missing `if`

alwa-siili replied:
Outdated

alwa-siili commented on templates/android/client/client.java.tpl:
```diff
@@ -0,0 +1,386 @@
+    private ClientHandler mClientHandler = new ClientHandler();
+
+    private final Map<Integer, Consumer<Bundle>> mpendingCalls = new ConcurrentHashMap<>();
+    AtomicInteger callIdsGetter = new AtomicInteger(0);
```
We use this `callIdsGetter` as a supplier of `callId`s.

Let's actually use the the `java.util.function.Supplier`:

```suggestion
    Supplier<Integer> callIdSupplier = new AtomicInteger(0)::getAndIncrement;
```

... then, use it as

```patch
-        int msgId =  callIdsGetter.getAndIncrement();
+        int msgId =  callIdSupplier.get();
```

alwa-siili commented on templates/android/client/client.java.tpl:
```diff
@@ -0,0 +1,510 @@
+{{- $interfacesToImport = (appendList $typesToImport $type) }}
+        {{- else }}
+{{- $typesToImport = (appendList $typesToImport $type) }}
+        {{- end }}
+    {{- end }}
+{{- end }}
+{{- range .Interface.Operations }}
+    {{- range .Params }}
+    {{- if and (not .Schema.Import)  (not .IsPrimitive)  }}
+{{- $type :=  Camel .Type }}
+        {{- if eq .KindType "interface" }}
+{{- $interfacesToImport = (appendList $typesToImport $type) }}
+        {{- else }}
+{{- $typesToImport = (appendList $typesToImport $type) }}
+        {{- end }}
+    {{- end }}
+    {{- end }}
+    {{- if and (and (not .Return.Schema.Import)  (not .Return.IsPrimitive))  (not .Return.IsVoid) }}
+{{- $type :=  Camel .Return.Type }}
+        {{- if eq .Return.KindType "interface" }}
+{{- $interfacesToImport = (appendList $typesToImport $type) }}
+        {{- else }}
+{{- $typesToImport = (appendList $typesToImport $type) }}
+        {{- end }}
+    {{- end }}
+{{- end }}
+{{- range .Interface.Signals }}
+    {{- range .Params }}
+    {{- if and (not .Schema.Import)  (not .IsPrimitive)  }}
+{{- $type :=  Camel .Type }}
+        {{- if eq .KindType "interface" }}
+{{- $interfacesToImport = (appendList $typesToImport $type) }}
+        {{- else }}
+{{- $typesToImport = (appendList $typesToImport $type) }}
+        {{- end }}
+    {{- end }}
+    {{- end }}
+{{- end }}
+{{- $typesToImport = unique $typesToImport }}
+{{- $interfacesToImport = unique $interfacesToImport }}
```
I'm not sure if finding all dependent include types should be a responsibility of a template itself. This seems more like something we should do on the go (apigear-cli) side. It would make the code much cleaner. And probably easeir to reuse, since template-java probably isn't the only template that may want to get all dependent types.

dorotaphanSiili replied:
I guess we've pick that up on some daily, I am not sure if a ticket should be made out of this.

alwa-siili commented on templates/android/messenger/enumparcelable.java.tpl:
```diff
@@ -0,0 +1,69 @@
+        for (int i = 0; i < enums.length; i++) {
+            result[i] = new {{Camel .Enum.Name }}Parcelable(enums[i]);
+        }
+        return result;
```
We could use a bit of functional programming here

```suggestion
        return Arrays.stream(enums)
            .map(e -> new {{Camel .Enum.Name }}Parcelable(e))
            .toArray({{Camel .Enum.Name }}Parcelable[]::new);
```

Same below

alwa-siili commented on templates/android/client/client.java.tpl:
```diff
@@ -0,0 +1,510 @@
+                case INIT:
+                {
+                    Bundle data = msg.getData();
+                    {{template "setClassLoaderIfNeeded" .Interface.Properties}}
+			        {{range .Interface.Properties}}
+                    {{template "getDataFromBundle" . }}
+				    on{{Camel .Name}}({{javaVar .}});
+			        {{- end}}
```
What does this `{{template "setClassLoaderIfNeeded" ...}}` do? A Bundle can have only one class loader set at a time. What is our intention here? Do we want to load all possible class loaders for properties and then iterate over all properties? This shouldn't work. Instead, we should load class loader specific to the property (if needed) just before getting the data from the bundle.

Actually, it could be done already in `getDataFromBundle` since we know, we should always do it for non-primitive types.

dorotaphanSiili replied:
class loaders already discussed and addressed

alwa-siili commented on templates/android/messenger/messages.java.tpl:
```diff
@@ -0,0 +1,48 @@
+
+{{ define "inc" }}{{ len (printf "%*s " . "") }}{{ end -}}
+
+public enum {{Camel .Interface.Name}}MessageType {
+    REGISTER_CLIENT(0),
+    UNREGISTER_CLIENT(1),
+    {{- $msgNum :=  1}}
+{{- range .Interface.Properties }}
+    {{- $msgNum =  len (printf "%*s " $msgNum "" ) }}
+    PROP_{{Camel .Name}}({{$msgNum}}),
+    {{- $msgNum =  len (printf "%*s " $msgNum "" ) }}
+    SET_{{Camel .Name}}({{$msgNum}}),
+{{- end }}
+{{- range .Interface.Signals }}
+    {{- $msgNum =  len (printf "%*s " $msgNum "" ) }}
+    SIG_{{Camel .Name}}({{$msgNum}}),
+{{- end }}
+{{- range .Interface.Operations }}
+    {{- $msgNum =  len (printf "%*s " $msgNum "" ) }}
+    RPC_{{Camel .Name}}Req({{$msgNum}}),
+    {{- $msgNum =  len (printf "%*s " $msgNum "" ) }}
+    RPC_{{Camel .Name}}Resp({{$msgNum}}),
+{{- end }}
```
If we're always creating message type enum starting from 0, couldn't we simply use enums without value and use built in `ordinal()`? Or do we always want to provide the value explicitly?

alwa-siili commented on templates/jnibridge/client/jnibridgeclient.java.tpl:
```diff
@@ -0,0 +1,140 @@
+    {{- range .Interface.Operations }}
+    private native void nativeOn{{Camel .Name}}Result({{if not .Return.IsVoid}}{{javaReturn "" .Return}} result, {{end}}String callId);
+    {{- end }}
+    private native void nativeIsReady(boolean isReady);
```
Consider changing this name to `nativeOnIsReady` to keep it consistent with the rest of the native "callbacks".

dorotaphanSiili replied:
not sure... we wanted apigear added functions to have different naming convention form ones added by user, not to clash with the user method in same interface

alwa-siili commented on rules.yaml:
```diff
@@ -3,7 +3,7 @@ features:
   - name: api
```
Please add an explanation why changing the `dot` to `camel` naming style is necessary in commit:

```
commit 43dc33ec43b3a41fbc67f6f8e925e878658b6ebc
Author: dorotaphanSiili <108282075+dorotaphanSiili@users.noreply.github.com>
Date:   Fri Aug 29 09:06:23 2025 +0200

    fix: change module name filter
```

As far as I remember, this is because android doesn't support modules with dots in the name. However, such important information shouldn't be passed from person to person. It should be kept in the commit message.

dorotaphanSiili committed:
fix(jni): fix unbind from client

When client unbinds correctly, there is no "confirmation" form service
side. It is assumed that if client want's to unbind it will always work.
Therefore internal cleaning needs to be done in this case.
Call to doCleanupForUnbinding was added for this case.
Now the client listners are informed that the client is not bound
anymore.
Also the cleanup for unbinding should not clean the client messenger,
this will not change through the client lifetime, and is created only
once, when the client is constructed. With this cleaned the client was
not able to use messenger in case of re-binding.

alwa-siili commented on templates/android/messenger/interfaceparcelable.java.tpl:
```diff
@@ -0,0 +1,170 @@
+
+    public I{{Camel .Interface.Name}} data;
+
+    public {{Camel .Interface.Name}}Parcelable(I{{Camel .Interface.Name}} data) {
+        this.data = data;
+    }
+
+    public I{{Camel .Interface.Name}} get{{Camel .Interface.Name}}()
+    {
+        return data;
+    }
```
Returning the reference without the "defensive copy" could be a bad idea. Instead, we should copy the `data` objects, just like we do in case of structures.

The only reasonable way to do this is to extend each interface with

```java
    I{{Camel .Interface.Name}} deepCopy();
```

... method, which must be overriden by each implementation to return deep copied self.

dorotaphanSiili replied:
already a ticket

dorotaphanSiili committed:
fix(jni): improve service backend cleanup

ServiceStarter is used to manage backend's lifetime, the app that uses it, decides with start and stop, when ServiceAdapter should have access to the backend.
It also mangaes when the ServiceAdapter should start and stop, but precise time of that is not guaranteed. One thing that it is asynchronous, second is that android service may be killed any time tha android lacks resource. The ServiceAdapter is an andorid service that forwards calls to and from the real backend, so it may be killed independently from the working app, the backend provider.
This commit only fixes the responsibility of cleanup.
The Service Starter is the one that should manage the backends lifetime and service adapters access to it.
This was fixed here. The service adapter no longer cleans those fields in the onDestroy function.
What is still missing is a mechanism to notify that andorid service was killed, and possibility to restart it. Such functionality should be added with separate commits.

dorotaphanSiili committed:
feat: handle silent service death

Service may get killed by android system while the application that starts it still works.
There is no notification this happened.
This commit removes the possible restart service.
The service was started with START_STICKY, which states that service may be scheduled to start at some point but it is not guaranteed.
It also does no give any notification it happend, and done by the android itself there would be no notification if service died agaiin.
So now we use START_NOT_STICKY, which means android will not try to restart it at any point by itself.
This commit adds the functionality of getting the notification (service killed, service started)
and forwarding them to the user of the ServiceStarter classes.
It will not schedule any restart, the user should implement restart logic. But has all the necessary tools now to detect such cases.
Commit introduces new class that collects the common functionality for both type of starters: for the stub and for the jni bridge.

dorotaphanSiili committed:
chores: rename factory to provider

The class purpose is to ensure that there is only one instance of the
backend, and provide this particular one always to the users.
The name was updated according to what it actually does.
